lib/Sema/CSApply.cpp

```diff
@@ -2514,10 +2514,91 @@ namespace {
+        if (!baseTyUnwrapped->getNominalOrBoundGenericNominal()) {
+          return;
+        }
+        
+        // Otherwise, continue digging
+        if (auto DSCE = dyn_cast<DotSyntaxCallExpr>(result)) {
+          auto calledValue = DSCE->getCalledValue();
+          auto isOptional = false;
+          Identifier memberName;
+          
+          // Try cast the assigned value to an enum case
+          //
+          // This will always succeed if the base is Optional<T> & the
+          // assigned case comes from Optional<T>
+          if (auto EED = dyn_cast<EnumElementDecl>(calledValue)) {
+            isOptional = EED->getParentEnum()->isOptionalDecl();
+            memberName = EED->getBaseName().getIdentifier();
+          }
+          
+          // Return if the enum case doesn't come from Optional<T>
+          if (!isOptional) {
+            return;
+          }
+          
+          // Look up the enum case in the unwrapped type to check if it exists
+          // as a member
+          auto baseTyNominalDecl = baseTyUnwrapped
+                                   ->getNominalOrBoundGenericNominal();
+          auto &tc = cs.getTypeChecker();
+          auto results = tc.lookupMember(baseTyNominalDecl->getModuleContext(),
+                                         baseTyUnwrapped,
+                                         memberName,
+                                         defaultMemberLookupOptions);
+          
+          for (LookupResultEntry result : results) {
+            auto member = result.getValueDecl();
+            // Lookup returned nothing or the member is an instance member,
+            // so return
+            if (!member && member->isInstanceMember()) {
+              break;
+            }
+            
+            // Return if the member is an enum case w/ assoc values, as we only
+            // care (for now) about cases with no assoc values (like none)
+            if (auto EED = dyn_cast<EnumElementDecl>(member)) {
+              if (EED->hasAssociatedValues()) {
+                break;
+              }
+            }
+            
+            // Emit a diagnostic with some fixits
+            auto baseTypeName = baseTy->getCanonicalType().getString();
+            auto baseTypeNameUnwrapped = baseTyUnwrapped->getString();
+            auto loc = DSCE->getLoc();
+            auto startLoc = DSCE->getStartLoc();
+            
+            tc.diagnose(loc, swift::diag::optional_ambiguous_case_ref,
+                        baseTypeName, baseTypeNameUnwrapped, memberName.str());
+            
+            tc.diagnose(loc, swift::diag::optional_fixit_ambiguous_case_ref)
+              .fixItInsert(startLoc, "Optional");
+            tc.diagnose(loc, swift::diag::type_fixit_optional_ambiguous_case_ref,
+                        baseTypeNameUnwrapped, memberName.str())
+              .fixItInsert(startLoc, baseTypeNameUnwrapped);
+            break;
```

Comment:
I think there's something off about your logic here, because all paths in this loop will hit a `break`. If you want to diagnose when any of the `LookupResultEntry`s match your criteria, the two `break`s in the `if` statements should be `continue`s; if you want to diagnose when all of the `LookupResultEntry`s match your criteria, then you probably need to declare a `bool` outside the loop, then change its value when you find one that doesn't match, and test it with `if` after the loop.